wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1194 @@
+        WOLFSSL_MSG("Response from iot-safe: too short");
+        return BAD_STATE_E;
+    }
+    rkey = search_tlv(resp, ret, IOTSAFE_TAG_ECC_KEY_FIELD);
+    if (!rkey) {
+        WOLFSSL_MSG("IoT safe Error in rkey response");
+        return MISSING_KEY;
+    }
+    ktype = search_tlv(rkey + 4, IOTSAFE_TAG_ECC_KEY_FIELD_SZ,
+            IOTSAFE_TAG_ECC_KEY_TYPE);
+    if (!ktype) {
+        WOLFSSL_MSG("IoT safe Error in ktype response");
+        return MISSING_KEY;
+    }
+    payload_str = search_tlv(ktype + 4, IOTSAFE_TAG_ECC_KEY_TYPE_SZ,
+            IOTSAFE_TAG_ECC_KEY_XY);
+    if (!payload_str) {
+        WOLFSSL_MSG("IoT safe: Error in payload response");
+        return MISSING_KEY;
+    }
+    XMEMSET(Qx, 0, sizeof(Qx));
+    XMEMSET(Qy, 0, sizeof(Qy));
+    ret = wc_ecc_init(key);
+    if (ret < 0) {
+        WOLFSSL_MSG("Cannot initialize ecc key to store IoTSafe public key");
+        return -1;
+    }
+    XSTRNCPY(Qx, payload_str + 6, IOTSAFE_ECC_KSIZE * 2);
+    XSTRNCPY(Qy, payload_str + 6 + IOTSAFE_ECC_KSIZE * 2, IOTSAFE_ECC_KSIZE * 2);
+
+    ret = wc_ecc_import_raw_ex(key, Qx, Qy, NULL, IOTSAFE_ECC_KTYPE);
+    if (ret < 0) {
+        WOLFSSL_MSG("Could not import raw key into ecc key");
+        return ret;
+    }
+    WOLFSSL_MSG("Get Public key: OK.\n");
+    return 0;
+}
+
+static int iotsafe_put_public_key(byte idx, ecc_key *key)
+{
+    char *resp;
+    int ret;
+    word32 qxlen=IOTSAFE_ECC_KSIZE, qylen=IOTSAFE_ECC_KSIZE;
+    byte ecc_pub_raw[IOTSAFE_TAG_ECC_KEY_FIELD_SZ] =  {
+        IOTSAFE_TAG_ECC_KEY_TYPE,
+        IOTSAFE_TAG_ECC_KEY_TYPE_SZ,
+        IOTSAFE_TAG_ECC_KEY_XY,
+        IOTSAFE_TAG_ECC_KEY_XY_SZ,
+        IOTSAFE_TAG_ECC_KEY_XY_HDR_BYTE
+    };
+
+    /* Export raw Qx, Qy values */
+    ret = wc_ecc_export_public_raw(key, ecc_pub_raw + 5, &qxlen,
+            ecc_pub_raw + 5 + IOTSAFE_ECC_KSIZE, &qylen);
+    if (ret != 0) {
+        WOLFSSL_MSG("IoT Safe: Could not export public key: Error");
+        return ret;
+    }
+
+    /* Put Public Init */
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_PUT_PUBLIC_INIT,
+            0, 0);
+    iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PUBKEY_ID, 1, &idx);
+    iotsafe_cmd_complete(csim_cmd);
+    if (expect_ok(csim_cmd, (word32)XSTRLEN(csim_cmd)) < 1) {
+        WOLFSSL_MSG("Unexpected reply when storing public key");
+        ret = WC_HW_E;
+    } else {
+        do {
+            ret = expect_ok("AT\r\n", 4);
+        } while (ret == 0);
+        if (ret > 0) {
+            /* Put Public Update */
+            iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_PUT_PUBLIC_UPDATE,
+                    IOTSAFE_DATA_LAST, 0);
+            iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_ECC_KEY_FIELD,
+                    IOTSAFE_TAG_ECC_KEY_FIELD_SZ, ecc_pub_raw);
+            iotsafe_cmd_complete(csim_cmd);
+            if (expect_csim_response(csim_cmd,
+                        (word32)XSTRLEN(csim_cmd), &resp) < 0) {
```

Comment:
Most other places error out on `<= 0`. Is this correct? This occurs also with `expect_ok` (lines 614, 651, 685, 717, 752). Is `0` acceptable in these locations?

Reply:
right, it should be > 0 as you said

Reply:
actually (after inspecting the IDE example test) some of the operations (e.g. 'update' commands) expect a zero-len reply after the last packet, so you'll find these checks still >= 0

Reply:
Ok. Just wanted to make sure that now that an error can be returned the checks are up to date.